File: code/jvm/im-http-api/src/main/kotlin/im/api/model/output/channel/ChannelOutputModel.kt
package im.api.model.output.channel

import im.api.model.output.users.UserOutputModel
import im.domain.channel.Channel

data class ChannelOutputModel(
    val id: Long,
    val name: String,
    val owner: UserOutputModel,
    val defaultRole: String,
    val isPublic: Boolean,
    val members: List<ChannelMemberOutputModel>,
    val createdAt: String,
) {
    companion object {
        fun fromDomain(channel: Channel) =
            ChannelOutputModel(
                id = channel.id.value,
                name = channel.name.value,
                owner = UserOutputModel.fromDomain(channel.owner),
                defaultRole = channel.defaultRole.name,
                isPublic = channel.isPublic,
                members = channel.members.map { ChannelMemberOutputModel.fromDomain(it.key, it.value) },
                createdAt = channel.createdAt.toString(),
            )
    }
}

File: code/js/src/Dto/output/channels/ChannelOutputModel.ts
import { UserOutputModel } from '../users/UserOutputModel';
import { ChannelMemberOutputModel } from './ChannelMemberOutputModel';
import { ChannelRole } from '../../../Domain/channel/ChannelRole';

/**
 * Channel member output model
 *
 * @param id The unique identifier of the member.
 * @param name The name of the member.
 * @param defaultRole The default role of a user in the channel.
 * @param owner The owner of the channel.
 * @param isPublic Indicates whether the channel is public.
 * @param members The members of the channel
 */
export interface ChannelOutputModel {
    id: number;
    name: string;
    defaultRole: ChannelRole;
    owner: UserOutputModel;
    isPublic: boolean;
    members: ChannelMemberOutputModel[];
}
